TypeCobol.Analysis/Cfg/ControlFlowGraphBuilder.cs

```diff
@@ -0,0 +1,2695 @@
+            var elseBlock = this.CurrentProgramCfgBuilder.CreateBlock(_else, true);
+            ctx.AddBranch(elseBlock);
+            //The new Current Block is the else block
+            this.CurrentProgramCfgBuilder.CurrentBasicBlock = elseBlock;
+        }
+
+        /// <summary>
+        /// Leave a Else instruction
+        /// </summary>
+        /// <param name="_else">The Else instruction</param>
+        protected virtual void LeaveElse(Else _else)
+        {
+
+        }
+
+        /// <summary>
+        /// Set whether or not the EVALUATE statement shall be translated using cascading IF-THEN-ELSE, false
+        /// otherwise.
+        /// </summary>
+        public bool UseEvaluateCascade
+        {
+            get;
+            set;
+        }
+
+        /// <summary>
+        /// Set whether or not the SEARCH statement shall be translated using cascading IF-THEN-ELSE, false
+        /// otherwise.
+        /// </summary>
+        public bool UseSearchCascade
+        {
+            get;
+            set;
+        }
+
+        /// <summary>
+        /// Enter an Evaluate statement
+        /// </summary>
+        /// <param name="evaluate">The Evaluate node</param>
+        protected virtual void EnterEvaluate(Evaluate evaluate)
+        {
+            System.Diagnostics.Debug.Assert(this.CurrentProgramCfgBuilder.CurrentBasicBlock != null);
+            MultiBranchContext ctx = new MultiBranchContext(this.CurrentProgramCfgBuilder, evaluate);
+            //Create a list of node of contextual When and WhenOther nodes.
+            ctx.ContextualData = new List<Node>();
+            if (this.CurrentProgramCfgBuilder.MultiBranchContextStack == null)
+            {
+                this.CurrentProgramCfgBuilder.MultiBranchContextStack = new Stack<MultiBranchContext>();
+            }
+            //Push and start the Evaluate context.
+            this.CurrentProgramCfgBuilder.MultiBranchContextStack.Push(ctx);
+            ctx.Start(this.CurrentProgramCfgBuilder.CurrentBasicBlock);
+            //Add the evaluate instruction to the current block
+            AddCurrentBlockNode(evaluate);
+        }
+
+        /// <summary>
+        /// Leave an Evaluate statement
+        /// </summary>
+        /// <param name="evaluate">The Evaluate node</param>
+        protected virtual void LeaveEvaluate(Evaluate evaluate)
+        {
+            System.Diagnostics.Debug.Assert(this.CurrentProgramCfgBuilder.MultiBranchContextStack != null);
+            System.Diagnostics.Debug.Assert(this.CurrentProgramCfgBuilder.MultiBranchContextStack.Count > 0);
+            MultiBranchContext ctx = this.CurrentProgramCfgBuilder.MultiBranchContextStack.Pop();
+            System.Diagnostics.Debug.Assert(ctx.Branches != null);
+
+            if (UseEvaluateCascade)
+            {   //Pop each MultiBranchContextStack instance till to the EVALUATE one
+                //and close each one.
+                while (ctx.Instruction == null)
```

Comment:
Shouldn't it be `while (ctx.Instruction != evaluate)` ?

Reply:
In fact we have a skype meeting about the explnation, here MultiBranchContext instance are those created by WhenConditionClause corresponding to When nodes, and here are not associated to a When node that has no CodeElement. Thus when the ctx.Instruction will be different than null, it will be the ctx corresponding to the evaluate.